params/governance_paramset.go

```diff
@@ -0,0 +1,419 @@
+
+func init() {
+	for name, key := range govParamNames {
+		govParamNamesReverse[key] = name
+	}
+}
+
+// GovParamSet is an immutable set of governance parameters
+// with various convenience getters.
+type GovParamSet struct {
+	// Items in canonical type.
+	// Only type checked and validated values will be stored.
+	items map[int]interface{}
+}
+
+func NewGovParamSet() *GovParamSet {
+	return &GovParamSet{
+		items: make(map[int]interface{}),
+	}
+}
+
+func NewGovParamSetStrMap(items map[string]interface{}) (*GovParamSet, error) {
+	p := NewGovParamSet()
+
+	for name, value := range items {
+		key, ok := govParamNames[name]
+		if !ok {
+			return nil, fmt.Errorf("Unknown governance param '%s'", name)
+		}
+		err := p.set(key, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func NewGovParamSetIntMap(items map[int]interface{}) (*GovParamSet, error) {
+	p := NewGovParamSet()
+
+	for key, value := range items {
+		err := p.set(key, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func NewGovParamSetChainConfig(config *ChainConfig) (*GovParamSet, error) {
+	items := make(map[int]interface{})
+	if config.Istanbul != nil {
+		items[Epoch] = config.Istanbul.Epoch
+		items[Policy] = config.Istanbul.ProposerPolicy
+		items[CommitteeSize] = config.Istanbul.SubGroupSize
+	}
+	items[UnitPrice] = config.UnitPrice
+	if config.Governance != nil {
+		items[GoverningNode] = config.Governance.GoverningNode
+		items[GovernanceMode] = config.Governance.GovernanceMode
+		if config.Governance.Reward != nil {
+			if config.Governance.Reward.MintingAmount != nil {
+				items[MintingAmount] = config.Governance.Reward.MintingAmount.String()
+			}
+			items[Ratio] = config.Governance.Reward.Ratio
+			items[UseGiniCoeff] = config.Governance.Reward.UseGiniCoeff
+			items[DeferredTxFee] = config.Governance.Reward.DeferredTxFee
+			items[StakeUpdateInterval] = config.Governance.Reward.StakingUpdateInterval
+			items[ProposerRefreshInterval] = config.Governance.Reward.ProposerUpdateInterval
+			if config.Governance.Reward.MinimumStake != nil {
+				items[MinimumStake] = config.Governance.Reward.MinimumStake.String()
+			}
+		}
+	}
+
+	return NewGovParamSetIntMap(items)
+}
+
+func (p *GovParamSet) set(key int, value interface{}) error {
+	ty, ok := govParamTypes[key]
+	if !ok {
+		return errors.New("Unknown governance param key")
+	}
+	parsed, ok := ty.ParseValue(value)
+	if !ok {
+		return errors.New("Malformed governance param value")
+	}
+	p.items[key] = parsed
+	return nil
+}
+
+func (p *GovParamSet) StrMap() map[string]interface{} {
+	m := map[string]interface{}{}
+	for key, value := range p.items {
+		m[govParamNamesReverse[key]] = value
+	}
+	return m
+}
+
+func (p *GovParamSet) IntMap() map[int]interface{} {
+	m := map[int]interface{}{}
+	for key, value := range p.items {
+		m[key] = value
+	}
+	return m
+}
+
+func (p *GovParamSet) Get(key int) (interface{}, bool) {
```

Comment:
By the way, why two types of `get` are provided? If you don't mind, can you write the comment about the difference?

Reply:
I added the comment about them in https://github.com/klaytn/klaytn/pull/1303/commits/e16a3626158d89e1ed87712e71f0c34edfc5f3cf.

`Get(key)` returns a boolean along with value. https://github.com/klaytn/klaytn/blob/v1.8.3/consensus/istanbul/backend/engine.go#L619 is a use case.

`MustGet(key)` always returns some value.
Main role of `MustGet(key)` is to implement nominal(noun-like) getters like `Epoch()`. 
Nominal getters should be the most popular way to read parameters.  https://github.com/klaytn/klaytn/blob/v1.8.3/consensus/istanbul/backend/engine.go#L230 is a use case.

As of now, I cannot determine if any one of them is unnecessary. I decided have both to serve both existing use cases.
When I replace all governance getters with GovParamSet (like in https://github.com/blukat29/klaytn/commit/147007c0), I will re-think about these two.

Feel free to comment. Thanks!

Reply:
Thanks for the comments